### PR TITLE
Esm push fixes

### PR DESCRIPTION
**Pull request by prateekbh**
**Bugfix**

**Did you add tests for your changes?**
Yes.

**Summary**
- With the addition of `--esm` we need to push module bundle instead of regular ones.
I also cleaned `template.html` a bit.
- This also adds the needed `crossorigin` attribute to preload the module script.

**Does this PR introduce a breaking change?**
No.

### CODE REVIEW

File: packages/cli/lib/lib/webpack/create-load-manifest.js [hunk 1]
```diff
@@ -1,14 +1,18 @@
...
+      } else if (/route-(.+)\.esm\.js/.test(filename) && isESMBuild) {
         scripts.push(filename);
       } else if (/chunk\.(.+)\.css$/.test(filename)) {
         styles.push(filename);
       } else if (/^bundle(.+)\.css$/.test(filename)) {
         mainCss = filename;
-      } else if (/^bundle(.+)\.js$/.test(filename)) {
+      } else if (!isESMBuild && /^bundle(.+)\.js$/.test(filename)) {
+        mainJs = filename;
+      } else if (isESMBuild && /bundle\.\w{5}\.esm\.js/.test(filename)) {
```
Comment (hassanbazzi) [inline packages/cli/lib/lib/webpack/create-load-manifest.js]: Such a crazy if statement we got here lol

Comment (prateekbh) [reply packages/cli/lib/lib/webpack/create-load-manifest.js]: heh.. i am not the best regexp maker out thr 😂

File: packages/cli/lib/resources/template.html [hunk 2]
```diff
@@ -27,8 +28,8 @@
 		}) %>
 		<% if (webpack.assets.filter(entry => entry.name.match(/bundle.\w{5}.esm.js$/)).length > 0) { %>
 			<!-- Fix for safari < 11 nomodule bug. TODO: Do the following only for safari.-->
-			<script>!function(){var e=document,t=e.createElement("script");if(!("noModule"in t)&&"onbeforeload"in t){var n=!1;e.addEventListener("beforeload",function(e){if(e.target===t)n=!0;else if(!e.target.hasAttribute("nomodule")||!n)return;e.preventDefault()},!0),t.type="module",t.src=".",e.head.appendChild(t),t.remove()}}();</script>
-			<script src="<%= htmlWebpackPlugin.files.publicPath %><%= webpack.assets.filter(entry => entry.name.match(/bundle.\w{5}.esm.js$/))[0].name %>" type="module"></script>
+			<script defer>!function(){var e=document,t=e.createElement("script");if(!("noModule"in t)&&"onbeforeload"in t){var n=!1;e.addEventListener("beforeload",function(e){if(e.target===t)n=!0;else if(!e.target.hasAttribute("nomodule")||!n)return;e.preventDefault()},!0),t.type="module",t.src=".",e.head.appendChild(t),t.remove()}}();</script>
```
Comment (rodush) [inline packages/cli/lib/resources/template.html]: `defer` does not make any sense for inline script, only to those with the `src` attribute. Recently found out for myself.

File: packages/cli/lib/lib/webpack/create-load-manifest.js [hunk 3]
```diff
@@ -24,18 +28,23 @@ module.exports = assets => {
...
     css = styles[idx];
     obj = Object.assign({}, defaults);
     obj[filename] = { type:'script', weight:0.9 };
     if (css) obj[css] = { type:'style', weight:0.9 };
-    path = filename.replace(/route-/, '/').replace(/\.chunk(\.\w+)?\.js$/, '').replace(/\/home/, '/');
-    manifest[path] = obj;
+    path = filename.replace(/route-/, '/')
+      .replace(/\.chunk(\.\w+)?\.js$/, '')
+      .replace(/\.chunk(\.\w+)?\.esm\.js$/, '')
+      .replace(/\/home/, '/');
```
Comment (lukeed) [inline packages/cli/lib/lib/webpack/create-load-manifest.js]: We already relied on `home.js` before (which we should probably fix), but by removing:

```js
manifest: {
  '/': defaults
}
```

...we're now _requiring_ a file named `routes/home.js` in order to populate a manifest entry for the `/` path.

File: packages/cli/lib/lib/webpack/create-load-manifest.js (hunk 1 above)
Comment (lukeed) [reply packages/cli/lib/lib/webpack/create-load-manifest.js]: For this block & the `route-` one above it, it probably cleaner to do something like this instead:

```js
let isESM = x => /\.esm\.js$/.test(x);
let isMatch = (x, y) => isESM(x) === y;

//---

      if (/route-/.test(filename)) {
        // both ESM & regular match here
        isMatch(filename, isESMBuild) && scripts.push(filename);
      } 
      // ...
      else if (/^bundle(.+)\.js$/.test(filename)) {
        // both bundles match here
        if (isMatch(filename, isESMBuild)) {
          mainJs = filename;
        }
      }
```

File: packages/cli/lib/lib/webpack/create-load-manifest.js (hunk 3 above)
Comment (prateekbh) [reply packages/cli/lib/lib/webpack/create-load-manifest.js]: oh nice catch, this was an intermittent thing when i was trying stuff.
I can put back 
```
manifest: {
  '/': defaults
}
```

Will that address your concern?

File: packages/cli/lib/resources/template.html (hunk 2 above)
Comment (prateekbh) [reply packages/cli/lib/resources/template.html]: addressed

File: packages/cli/lib/lib/webpack/create-load-manifest.js (hunk 3 above)
Comment (lukeed) [reply packages/cli/lib/lib/webpack/create-load-manifest.js]: I _think_ so, tho I'm not 100% sure

File: packages/cli/lib/lib/webpack/create-load-manifest.js (hunk 1 above)
Comment (prateekbh) [reply packages/cli/lib/lib/webpack/create-load-manifest.js]: done. Thanks for the suggestion

File: packages/cli/lib/resources/template.html [hunk 4]
```diff
@@ -10,9 +10,10 @@
 		<% if (htmlWebpackPlugin.options.manifest.theme_color) { %>
 			<meta name="theme-color" content="<%= htmlWebpackPlugin.options.manifest.theme_color %>">
 		<% } %>
-		<% for (const file in htmlWebpackPlugin.options.createLoadManifest(compilation.assets)[htmlWebpackPlugin.options.url]) { %>
-			<% if (file) { %>
-				<link rel="preload" href="<%= htmlWebpackPlugin.files.publicPath + file %>" as="<%= file.match(/\.css$/)?'style':'script' %>" <%= file.match(/\.esm\.js$/)?'crossorigin="anonymous"':'' %>>
+		<% for (const file in htmlWebpackPlugin.options.createLoadManifest(compilation.assets, htmlWebpackPlugin.options.config.esm)[htmlWebpackPlugin.options.url]) { %>
+			<% if (file && file.match(/\.(css|js)$/)) { %>
+				<!-- crossorigin for main bundle as that is loaded from <script type=module tag, other lazy loaded bundles are from webpack so its not needed -->
```
Comment (developit) [inline packages/cli/lib/resources/template.html]: probably can skip the comment or put it in `<% /* these gross tags */ %>`

**Comment by developit** [comment]
Quick question: is this going to push/preload both the esm and non-esm bundle?

**Comment by prateekbh** [comment]
@developit no it will push only esm bundles

**Comment by ForsakenHarmony** [comment]
How many browsers have push support but not module support?

**Comment by prateekbh** [comment]
@forsakenharmony, there are a good bunch of such browsers. But all last 3 version support both IIRC

**Comment by prateekbh** [comment]
Is this good to merge?

File: packages/cli/lib/lib/webpack/create-load-manifest.js [hunk 5]
```diff
@@ -1,15 +1,22 @@
-module.exports = assets => {
+const isESM = filename => /\.esm\.js$/.test(filename);
```
Comment (developit) [inline packages/cli/lib/lib/webpack/create-load-manifest.js]: Not entirely scoped to this PR, but would you be open to using `.mjs`?

File: packages/cli/lib/lib/webpack/push-manifest.js [hunk 6]
```diff
@@ -1,9 +1,12 @@
 const createLoadManifest = require('./create-load-manifest');
 
 module.exports = class PushManifestPlugin {
+  constructor(env = {}) {
```
Comment (developit) [inline packages/cli/lib/lib/webpack/push-manifest.js]: Spacing seems off here (or it's my phone). Also the underscore likely trips our lint rules.

File: packages/cli/lib/lib/webpack/create-load-manifest.js [hunk 7]
```diff
@@ -34,7 +41,10 @@ module.exports = assets => {
     obj = Object.assign({}, defaults);
     obj[filename] = { type:'script', weight:0.9 };
     if (css) obj[css] = { type:'style', weight:0.9 };
-    path = filename.replace(/route-/, '/').replace(/\.chunk(\.\w+)?\.js$/, '').replace(/\/home/, '/');
+    path = filename.replace(/route-/, '/')
+      .replace(/\.chunk(\.\w+)?\.js$/, '')
```
Comment (developit) [inline packages/cli/lib/lib/webpack/create-load-manifest.js]:
```
.replace(/\.chunk(\.\w+)?(\.esm)?\.js$/, '')
```

**Comment by developit** [comment]
I didn't run this locally yet to verify but otherwise just the above nitpicks. 

Also I'm not sure if it's elsewhere but it might be nice to expose `BUNDLE_FORMAT="module"` via `env` in case folks need to change behaviour in response to the format in use.

File: packages/cli/lib/lib/webpack/create-load-manifest.js (hunk 5 above)
Comment (prateekbh) [reply packages/cli/lib/lib/webpack/create-load-manifest.js]: I am open to do this in another PR. IIRC webpack's uglify only runs on .just files by default so I switched.

File: packages/cli/lib/lib/webpack/push-manifest.js (hunk 6 above)
Comment (prateekbh) [reply packages/cli/lib/lib/webpack/push-manifest.js]: done.

File: packages/cli/lib/resources/template.html (hunk 4 above)
Comment (prateekbh) [reply packages/cli/lib/resources/template.html]: done

File: packages/cli/lib/lib/webpack/create-load-manifest.js (hunk 7 above)
Comment (prateekbh) [reply packages/cli/lib/lib/webpack/create-load-manifest.js]: this one isn't working hence keeping as is

Comment (prateekbh) [reply packages/cli/lib/lib/webpack/create-load-manifest.js]: nevermind... fixed it

**Comment by prateekbh** [comment]
@developit fixed the nits.
> Also I'm not sure if it's elsewhere but it might be nice to expose BUNDLE_FORMAT="module" via env in case folks need to change behaviour in response to the format in use.

This is exactly what [this](https://github.com/developit/preact-cli/blob/next/packages/cli/lib/lib/webpack/webpack-client-config.js#L112) does, just a different var name.

**Comment by developit** [comment]
Ah!

**Pull request merged**